packages/flutter/lib/src/material/chip.dart

```diff
@@ -103,81 +111,658 @@ class Chip extends StatelessWidget {
+  void insertChildRenderObject(RenderObject child, dynamic slotValue) {
+    assert(child is RenderBox);
+    assert(slotValue is _ChipSlot);
+    final _ChipSlot slot = slotValue;
+    _updateRenderObject(child, slot);
+    assert(renderObject.childToSlot.keys.contains(child));
+    assert(renderObject.slotToChild.keys.contains(slot));
+  }
+
+  @override
+  void removeChildRenderObject(RenderObject child) {
+    assert(child is RenderBox);
+    assert(renderObject.childToSlot.keys.contains(child));
+    _updateRenderObject(null, renderObject.childToSlot[child]);
+    assert(!renderObject.childToSlot.keys.contains(child));
+    assert(!renderObject.slotToChild.keys.contains(slot));
+  }
+
+  @override
+  void moveChildRenderObject(RenderObject child, dynamic slotValue) {
+    assert(false, 'not reachable');
+  }
+}
+
+class _ChipRenderTheme {
+  const _ChipRenderTheme({
+    @required this.avatar,
+    @required this.label,
+    @required this.deleteIcon,
+    @required this.container,
+    @required this.padding,
+    @required this.labelPadding,
+  });
+
+  final Widget avatar;
+  final Widget label;
+  final Widget deleteIcon;
+  final Widget container;
+  final EdgeInsetsDirectional padding;
+  final EdgeInsetsDirectional labelPadding;
+
+  @override
+  bool operator ==(dynamic other) {
+    if (identical(this, other)) {
+      return true;
+    }
+    if (other.runtimeType != runtimeType) {
+      return false;
+    }
+    final _ChipRenderTheme typedOther = other;
+    return typedOther.avatar == avatar &&
+        typedOther.label == label &&
+        typedOther.deleteIcon == deleteIcon &&
+        typedOther.container == container &&
+        typedOther.padding == padding &&
+        typedOther.labelPadding == labelPadding;
+  }
+
+  @override
+  int get hashCode {
+    return hashValues(
+      avatar,
+      label,
+      deleteIcon,
+      container,
+      padding,
+      labelPadding,
+    );
+  }
+}
+
+class _RenderChip extends RenderBox {
+  _RenderChip({
+    @required _ChipRenderTheme theme,
+    @required TextDirection textDirection,
+    this.onDeleted,
+  })  : assert(theme != null),
+        assert(textDirection != null),
+        _theme = theme,
+        _textDirection = textDirection {
+    _tap = new TapGestureRecognizer(debugOwner: this)
+      ..onTapDown = _handleTapDown
+      ..onTap = _handleTap;
+  }
+
+  // Set this to true to have outlines of the tap targets drawn over
+  // the chip.  This should never be checked in while set to 'true'.
+  static const bool _debugShowTapTargetOutlines = false;
+  static const EdgeInsetsDirectional _iconPadding = const EdgeInsetsDirectional.fromSTEB(
+    _edgePadding,
+    _edgePadding,
+    _edgePadding,
+    _edgePadding,
+  );
+
+  final Map<_ChipSlot, RenderBox> slotToChild = <_ChipSlot, RenderBox>{};
+  final Map<RenderBox, _ChipSlot> childToSlot = <RenderBox, _ChipSlot>{};
+
+  TapGestureRecognizer _tap;
+
+  VoidCallback onDeleted;
+  Rect _deleteButtonRegion;
```

Comment:
Maybe this should be called _deleteButtonRect.

Reply:
Done.